### ISSUE TITLE
midi functions

### ISSUE THREAD

**Issue by giuliomoro** [libpd/libpd#133]
Hi, I tried to send messages to `int libpd_midibyte(int port, int value)`. Such messages are received only by [midiin] objects but are note parsed and forwarded to, e.g.: [notein] [ctlin] [pgmin] objects.

In Pd, all the messages coming out of [midin] are also internally routed to the appropriate [notein] [ctlin] [pgmin] objects. Similarly, I would expect that sending a raw midibyte to `libpd_midibyte` should produce outputs, as appropriate, from all the  [notein] [ctlin] [pgmin]  objects in the patch.

**Comment by danomatika** [comment]
libpd doesn't do any internal midi byte parsing which is why this doesn't happen.

Pd vanilla accomplishes this as each byte gets passed on to the Portmidi library which parses the bytes and forwards them to the appropriate receiver which is, internally, #notein, #ctlin, #pgmin, etc.

We could add byte parsing but the thought is: that is best left to whichever MIDI library or implementation the libpd user decides on. This is the same approach we use for the audio driver.

If you are in control of the sending, why not use `libpd_noteon()`, `libpd_controlchange()`, & `libpd_programchange()` instead?

**Comment by danomatika** [comment]
Also, you can parse raw midi bytes within a pd patch, if required.

See rc-notein, rc-ctlin, & rc-pgmin in https://github.com/danomatika/rc-patches/tree/master/deprecated/rc-patches.

**Comment by giuliomoro** [comment]
thanks for your answer.
I surely can send  to `libpd_noteon()`, `libpd_controlchange()`, `libpd_programchange()` , just that this way any patch that was relying on [midiin] (assuming any exists) will need to be re-written accordingly.

I was actually just checking what the reason is behind this and how to work with it. Sounds like, as I am not using portmidi but my own parser, the right way to do this is to send to `libpd_noteon()`, `libpd_controlchange()`, `libpd_programchange()`  AND `libpd_midibyte()`

**Comment by danomatika** [comment]
If you rely on using [midiin] as well, that's probably the easiest thing to do, yeah. The number of patches that use [midi] should be relatively low, however.

**Comment by giuliomoro** [comment]
good, thanks.

**Comment by danomatika** [comment]
If you are parsing a byte stream, you can of course forward each byte to midi as well as send the detected events. That what I do in PdParty: <https://github.com/danomatika/PdParty/blob/master/src/modules/Midi.m>

(Also, I'll add [midiin] support to PdParty as this fact hadn't occurred to me!)

**Comment by danomatika** [comment]
1 more thing: You can also avoid using the `libpd_midibyte()` function with it's weird port argument by sending directly to the internal #midiin receiver, I believe. That's worth a try.

**Comment by tkirshboim** [comment]
@danomatika Is this difference between libpd and pd-vanilla in the `midiin` behavior documented anywhere? I couldn't find it in the [midi part in the wiki](https://github.com/libpd/libpd/wiki/libpd#midi-support-in-libpd) and recently @monkeyswarm brought this up as an [issue in the pd-for-android project](https://github.com/libpd/pd-for-android/issues/48).
Also, regarding that issue - is it correct that `midiin` in pd-for-ios behaves like the pd-vanilla version? That could add up to some more confusion then..

**Comment by danomatika** [comment]
>Is this difference between libpd and pd-vanilla in the midiin behavior documented anywhere? I couldn't find it in the midi part in the wiki and recently @monkeyswarm brought this up as an issue in the pd-for-android project.

Feel free to add it to the wiki. It's a wiki after all...

>is it correct that midiin in pd-for-ios behaves like the pd-vanilla version? 

No. pd-for-ios, as with all other language/environment wrappers, does not include raw midi parsing.

**Comment by tkirshboim** [comment]
> Feel free to add it to the wiki. It's a wiki after all...
Yes, will do. I wanted to be sure first that this is the expected behavior. That's for the clarification.

**Comment by danomatika** [comment]
The real fix would be to construct & send raw midi messages for each message type explicitly aka within libpd_noteon, etc but I'd ask @nettoyeurny what he thinks first.

**Comment by tkirshboim** [comment]
I added a section on this to [the wiki page](https://github.com/libpd/libpd/wiki/libpd#difference-in-midiin-functionality-compared-to-pd-vanilla). Feel free to revise it if something there is incorrect.

**Comment by danomatika** [comment]
I'm thinking of adding a midi parsing layer which would provide this functionality as an option util. Basically it would work like the print concatenation layer with a custom hook for parsing the bytes from from pd and wrappers for the send functions to also send to the [midiin] object.
